### PR TITLE
Iox #932 handle nullptr callbacks in waitset

### LINKED ISSUE #932: Handle nullptr callbacks correctly in WaitSet and Listener
## Brief description

When checking the WaitSet and Listener tests with ASAN on gcc an error is thrown that a nullptr is de-referenced when a nullptr is passed as callback function to functions like `iox_ws_attach_subscriber_event` or `iox_listener_attach_subscriber_event` in the C-Binding.

In the WaitSet it should be possible for the users to use nullptr callback arguments if they want to wait in a loop.

Reference Links:
https://github.com/eclipse-iceoryx/iceoryx/pull/918#discussion_r716790474
https://github.com/eclipse-iceoryx/iceoryx/pull/918#discussion_r718865631

- [ ] Enable in CI the ASAN check with latest gcc compiler
- [ ] Refactor WaitSet and Listener to accept (or disallow) nullptr arguments
- [ ] Test explicitly against this

### PR DESCRIPTION
**Pull request by dkroenke**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #932

### CODE REVIEW

File: iceoryx_binding_c/test/moduletests/test_notification_info.cpp [hunk 1]
```diff
@@ -91,6 +91,9 @@ class iox_notification_info_test : public Test
         return m_memoryManager.getChunk(chunkSettings);
     }
 
+    static void triggerCallback(iox_sub_t sub IOX_MAYBE_UNUSED)
```
Comment (elfenpiff) [inline iceoryx_binding_c/test/moduletests/test_notification_info.cpp:94]: You can also get remove the variable name `sub` to indicate to the compiler that you do not use the function argument. This is less code and actually the default way.

**Review by elfenpiff — APPROVED** [review]
Since you opened Pandora's box.

The handling of the WaitSet is correct and I would merge this PR but there is something missing. It does not make any sense to provide a nullptr callback to a listener since it is purely event/callback based therefore you always have to attach a callback. A little oversight from my side but now you have to pay the price ;)

Could you create an additional PR and do the following:
1. listener.hpp
   1. Add enum value `ListenerError::EMPTY_EVENT_CALLBACK`
   2. Add check in `Listener::attachEvent` that returns the new enum value when someone tries to add an `eventCallback` with a `nullptr`  set for `m_callback`.
   3. Add tests for this new behavior
 2. binding_c: enums.h
    1. Add the value `ListenerResult_EMPTY_EVENT_CALLBACK`
    2. Add this value in the translation function in `cpp2c_enum_translation.cpp`
    3. Add a test for this additional translation
 3. binding_c: listener.h / c_listener.cpp
   1. Perform the adjustments like in the Waitset, you do not need to perform error checking on this level since the C++ class is doing this already for you. Just forward the arguments and do not use `createNotificationCallback`, if `m_callback` is `nullptr` `ListenerError::EMPTY_EVENT_CALLBACK` is returned.
   2. Add a test to verify the `NULL` error handling in C

File: iceoryx_binding_c/test/moduletests/test_notification_info.cpp [hunk 2]
```diff
@@ -142,7 +145,7 @@ TEST_F(iox_notification_info_test, notificationOriginIsUserTriggerPointerWhenIts
 
 TEST_F(iox_notification_info_test, notificationOriginIsSubscriberPointerWhenItsOriginatingFromThemStateBased)
 {
-    iox_ws_attach_subscriber_state(&m_waitSet, m_subscriberHandle, SubscriberState_HAS_DATA, 587U, NULL);
+    iox_ws_attach_subscriber_state(&m_waitSet, m_subscriberHandle, SubscriberState_HAS_DATA, 587U, triggerCallback);
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_notification_info.cpp:148]: Due to the changes in `c_wait set.cpp` we could here still use `NULL`, right?
No changes needed.

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 3]
```diff
@@ -142,12 +142,28 @@ TEST_F(iox_ws_test, SizeIsZeroWhenConstructed)
...
+}
+
+TEST_F(iox_ws_test, SizeIsOneWhenOneUserTriggerWithNullptrCallBackIsAttached)
+{
+    EXPECT_EQ(iox_ws_attach_user_trigger_event(m_sut, m_userTrigger[0U], 0U, NULL),
+              iox_WaitSetResult::WaitSetResult_SUCCESS);
+    EXPECT_EQ(iox_ws_size(m_sut), 1U);
+}
+
+TEST_F(iox_ws_test, SizeIsOneWhenOneSubscriberStateWithNullptrCallBackIsAttached)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:166]: Please add a similar test for a subscriber event with a nullptr callback.

File: iceoryx_binding_c/test/moduletests/test_notification_info.cpp (hunk 1 above)
Comment (dkroenke) [reply iceoryx_binding_c/test/moduletests/test_notification_info.cpp:94]: I think the `unused` attribute is not applicable to types only

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp (hunk 3 above)
Comment (dkroenke) [reply iceoryx_binding_c/test/moduletests/test_wait_set.cpp:166]: @FerdinandSpitzschnueffler done

File: iceoryx_binding_c/test/moduletests/test_notification_info.cpp (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_binding_c/test/moduletests/test_notification_info.cpp:94]: I think you can remove `IOX_MAYBE_UNUSED` as well and just write `static void triggerCallback(iox_sub_t)`.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
